### PR TITLE
Rework yaml2proto to accept multiple YAML files

### PR DESCRIPTION
**Pull request by ixdy**
Also add `-print-text` option to help with manual verification.

This ended up being less trivial than I'd hoped because I needed to add logic to handle updating the proto multiple times. Hopefully it all makes sense?

I'm guessing `testgrid/jenkins_verify/jenkins_validate.go` will need to be made aware of multiple YAML files as well, but this PR is already getting large enough as-is.

### CODE REVIEW

File: testgrid/config/yaml2proto/yaml2proto.go [hunk 1]
```diff
@@ -84,54 +89,110 @@ func ValidateDashboardtab(currentTab *config.DashboardTab, defaultTab *config.Da
...
-	// Unmarshal yaml to config
-	curConfig := &config.Configuration{}
-	defaultConfig := &config.DefaultConfiguration{}
-	err := Unmarshal(yamlData, &curConfig)
+// updateDefaults reads any default configuration from yamlData and updates the
+// defaultConfig in c.
+//
+// Returns an error if the defaultConfig remains unset.
+func (c *Config) updateDefaults(yamlData []byte) error {
+	new := &config.DefaultConfiguration{}
```
Comment (krzyzacy) [inline testgrid/config/yaml2proto/yaml2proto.go]: ah, new is a valid var name in go!

Comment (ixdy) [inline testgrid/config/yaml2proto/yaml2proto.go]: renamed.

**Comment by ixdy** [comment]
I hacked a `MarshalText` into the current yaml2proto:
```diff
diff --git a/testgrid/config/yaml2proto/yaml2proto.go b/testgrid/config/yaml2proto/yaml2proto.go
index 6ff463d6..44f7f37d 100644
--- a/testgrid/config/yaml2proto/yaml2proto.go
+++ b/testgrid/config/yaml2proto/yaml2proto.go
@@ -20,6 +20,7 @@ import (
        "errors"
        "fmt"
        "gopkg.in/yaml.v2"
+       "os"
 
        "github.com/golang/protobuf/proto"
        "k8s.io/test-infra/testgrid/config/pb"
@@ -132,6 +133,7 @@ func Yaml2Proto(yamlData []byte) ([]byte, error) {
                }
        }
 
+       proto.MarshalText(os.Stdout, curConfig)
        // Marshal config to protobuf
        return proto.Marshal(curConfig)
 }
```

The output from this matches the -print-text output from my changes.

**Comment by krzyzacy** [comment]
@k8s-bot I can LGTM!
@k8s-bot tell me a joke

File: testgrid/config/main.go [hunk 2]
```diff
@@ -17,40 +17,63 @@ limitations under the License.
...
 
-	protobufData, err := yaml2proto.Yaml2Proto(yamlData)
-	if err != nil {
-		fmt.Printf("Yaml2Proto Error : %v\n", err)
-		os.Exit(-1)
+	var c yaml2proto.Config
+	for _, file := range yamlFiles {
+		b, err := ioutil.ReadFile(file)
+		if err != nil {
+			errExit("IO Error : Cannot Read File %v\n", file)
```
Comment (spxtr) [inline testgrid/config/main.go]: `errExit("IO Error : Cannot Read File %s : %v\n", file, err)`

File: testgrid/config/main.go [hunk 3]
```diff
@@ -17,40 +17,63 @@ limitations under the License.
...
+		if err := c.MarshalText(os.Stdout); err != nil {
+			errExit("err printing proto: %v", err)
+		}
+	} else {
+		b, err := c.MarshalBytes()
+		if err != nil {
+			errExit("err encoding proto: %v", err)
+		}
+		err = ioutil.WriteFile(*outputPath, b, 0644)
+		if err != nil {
```
Comment (spxtr) [inline testgrid/config/main.go]: nit: combine with the line above

File: testgrid/config/yaml2proto/yaml2proto.go [hunk 4]
```diff
@@ -84,54 +89,110 @@ func ValidateDashboardtab(currentTab *config.DashboardTab, defaultTab *config.Da
...
-	curConfig := &config.Configuration{}
-	defaultConfig := &config.DefaultConfiguration{}
-	err := Unmarshal(yamlData, &curConfig)
+// updateDefaults reads any default configuration from yamlData and updates the
+// defaultConfig in c.
+//
+// Returns an error if the defaultConfig remains unset.
+func (c *Config) updateDefaults(yamlData []byte) error {
+	newDefaults := &config.DefaultConfiguration{}
+	err := yaml.Unmarshal(yamlData, newDefaults)
```
Comment (spxtr) [inline testgrid/config/yaml2proto/yaml2proto.go]: nit: combine with the line below

File: testgrid/config/yaml2proto/yaml2proto_test.go [hunk 5]
```diff
@@ -57,48 +55,82 @@ dashboards:
...
+  name: default`,
+			expectedError: nil,
+		},
+	}
+
+	for index, test := range tests {
+		c := Config{}
+		err := c.Update([]byte(test.yaml))
+		if (err == nil && test.expectedError == nil) ||
+			(err != nil && test.expectedError != nil && err.Error() == test.expectedError.Error()) {
```
Comment (spxtr) [inline testgrid/config/yaml2proto/yaml2proto_test.go]: Never check error strings in tests. Instead, check with type assertions.

Comment (ixdy) [inline testgrid/config/yaml2proto/yaml2proto_test.go]: Making me fix stuff I didn't write. Fine! :)

**Comment by ixdy** [comment]
Thanks for the review. PTAL.

File: testgrid/config/yaml2proto/yaml2proto_test.go (hunk 5 above)
Comment (krzyzacy) [inline testgrid/config/yaml2proto/yaml2proto_test.go]: ah learned! Thanks!

**Pull request merged**